style.md

```diff
@@ -887,112 +893,81 @@ if err := foo.Open(); err != nil {
+<thead><tr><th>No error matching</th><th>Error matching</th></tr></thead>
 <tbody>
 <tr><td>
 
 ```go
-func open(file string) error {
-  return fmt.Errorf("file %q not found", file)
-}
-
-func use() {
-  if err := open("testfile.txt"); err != nil {
-    if strings.Contains(err.Error(), "not found") {
-      // handle
-    } else {
-      panic("unknown error")
-    }
-  }
-}
-```
-
-</td><td>
-
-```go
-type errNotFound struct {
-  file string
-}
+// package foo
 
-func (e errNotFound) Error() string {
-  return fmt.Sprintf("file %q not found", e.file)
+func Open(file string) error {
+  return fmt.Errorf("file %q not found", file)
 }
 
-func open(file string) error {
-  return errNotFound{file: file}
-}
+// package bar
 
-func use() {
-  if err := open("testfile.txt"); err != nil {
-    if _, ok := err.(errNotFound); ok {
-      // handle
-    } else {
-      panic("unknown error")
-    }
-  }
+if err := foo.Open("testfile.txt"); err != nil {
+  // Can't handle the error.
+  panic("unknown error")
 }
 ```
 
-</td></tr>
-</tbody></table>
-
-Be careful with exporting custom error types directly since they become part of
-the public API of the package. It is preferable to expose matcher functions to
-check the error instead.
+</td><td>
 
 ```go
 // package foo
 
-type errNotFound struct {
-  file string
-}
-
-func (e errNotFound) Error() string {
-  return fmt.Sprintf("file %q not found", e.file)
+type NotFoundError struct {
+  File string
 }
 
-func IsNotFoundError(err error) bool {
-  _, ok := err.(errNotFound)
-  return ok
+func (e NotFoundError) Error() string {
+  return fmt.Sprintf("file %q not found", e.File)
 }
 
 func Open(file string) error {
-  return errNotFound{file: file}
+  return NotFoundError{File: file}
 }
 
+
 // package bar
 
-if err := foo.Open("foo"); err != nil {
-  if foo.IsNotFoundError(err) {
-    // handle
+if err := foo.Open("testfile.txt"); err != nil {
+  var notFound NotFoundError
+  if errors.As(err, &notFound) {
+    // handle the error
   } else {
     panic("unknown error")
   }
 }
 ```
 
-<!-- TODO: Exposing the information to callers with accessor functions. -->
+</td></tr>
+</tbody></table>
 
-### Error Wrapping
+Note that if you export error variables or types from a package,
+they will become part of the public API of the package.
+
+#### Error Wrapping
 
 There are three main options for propagating errors if a call fails:
 
 - Return the original error if there is no additional context to add and you
   want to maintain the original error type.
-- Add context using [`"pkg/errors".Wrap`] so that the error message provides
-  more context and [`"pkg/errors".Cause`] can be used to extract the original
-  error.
-- Use [`fmt.Errorf`] if the callers do not need to detect or handle that
-  specific error case.
+- Add context using `fmt.Errorf` with a `%w` verb if you want callers to
+  be able to match against or extract the original error.
+  If you do this, note that this is now part of your public API.
```

Comment:
Not unless you say it is :)

Reply:
Yes, but I mean, Hyrum's Law applies. If you were exposing it before, and users rely on having the underlying cause available, changing that is a breaking change regardless of whether you documented it. But I see your point. I'll see if there's a way to reword this a little so authors are deliberate about the choice of using %v vs %w.

Reply:
Hyrum's Law cannot dictate API compatibility. If it did, then literally any user-visible change is a breaking change, which is facially incoherent.

Reply:
Fair point. I guess I'm trying to find a way to suggest that the choice should be deliberate. I'm trying to work this section up a bit more.

Reply:
IME it's quite difficult to have enough knowledge/context to be able to make a deliberate choice on stuff like this! 

$0.02 — I've always suggested to use `%w` by default, and `%v` only when you explicitly want to hide information.

Reply:
Also reasonable. Internal discussion about this one yielded a similar result.
I think I've found a good balance of "use `%w` but be deliberate about it when necessary".
PR incoming.

Reply:
134

Reply:
https://github.com/uber-go/guide/pull/135